compiler/wasm/src/compile.rs

```diff
@@ -85,32 +96,44 @@ pub fn compile(
     }
 }
 
-fn add_noir_lib(context: &mut Context, library_name: &str) {
-    let path_to_lib = Path::new(&library_name).join("lib.nr");
-    let library_crate_id = prepare_dependency(context, &path_to_lib);
+fn process_dependency_graph(context: &mut Context, dependency_graph: DependencyGraph) {
+    let mut crate_names: HashMap<String, CrateId> = HashMap::new();
 
-    add_dep(context, *context.root_crate_id(), library_crate_id, library_name.parse().unwrap());
-
-    // TODO: Remove this code that attaches every crate to every other crate as a dependency
-    let root_crate_id = context.root_crate_id();
-    let stdlib_crate_id = context.stdlib_crate_id();
-    let other_crate_ids: Vec<_> = context
-        .crate_graph
-        .iter_keys()
-        .filter(|crate_id| {
-            // We don't want to attach this crate to itself or stdlib, nor re-attach it to the root crate
-            crate_id != &library_crate_id
-                && crate_id != root_crate_id
-                && crate_id != stdlib_crate_id
-        })
-        .collect();
+    // register libraries
+    for lib in &dependency_graph.root_dependencies {
+        let crate_id = add_noir_lib(context, &lib);
+        crate_names.insert(lib.to_string(), crate_id);
+    }
 
-    for crate_id in other_crate_ids {
-        context
-            .crate_graph
-            .add_dep(crate_id, library_name.parse().unwrap(), library_crate_id)
-            .unwrap_or_else(|_| panic!("ICE: Cyclic error triggered by {library_name} library"));
+    // make root crate as depending on it
+    for lib in &dependency_graph.root_dependencies {
+        let crate_id = crate_names.get(lib.as_str()).unwrap().clone();
+        let root_crate_id = context.root_crate_id().clone();
+        add_dep(context, root_crate_id, crate_id, lib.parse().unwrap());
     }
+
+    for (lib_name, dependencies) in dependency_graph.library_dependencies {
+        let crate_id = crate_names.get(lib_name.as_str()).unwrap().clone();
```

Comment:
We shouldn't unwrap here as it's not guaranteed that the user will have included this library in the `root_dependencies` vector.